KIGEN: Bootstrap kigen project

Jeff-Ciesielski opened the pull request:
This is a rather sizable PR, but it's really split into three parts:

1) The readme
2) The actual guts of kigen
3) The tests + collateral

Commits are logically separate and go to state of mind, but honestly I'd recommend reviewing the end result as there was a lot of shifting around as I felt it out.

This has a reasonably decent testing suite, and the CLI functionality was tested locally.
For an in-depth explanation of usecases, see the README (or ping me)

@keyme/robotics
 
Still to-do: 
* setup.py file
* publishing to pypi

Jeff-Ciesielski committed:
Initial commit

Extraction core is there, so we can grab autogen blocks out of strings
now.  Next step is to add generation, then the plugin system (for
loading templates from folders) and a CLI.

Jeff-Ciesielski committed:
Add basic expansion support + test

Jeff-Ciesielski committed:
Convert test to use actual directory

Jeff-Ciesielski committed:
Add module loading and test

Jeff-Ciesielski committed:
Add block/cmd to string helper functions and tests

Jeff-Ciesielski committed:
Refactor module extraction. Add recombination + tests

Jeff-Ciesielski committed:
Add full file expansion and test

IT WORKS!

Jeff-Ciesielski committed:
Add CLI and read/write file capablities

Jeff-Ciesielski committed:
Add documentation and minor cleanup

Jeff-Ciesielski committed:
Add readme

zsalzbank commented on README.md:
```diff
@@ -0,0 +1,147 @@
+# KIGen: Inline templating for source code (or whatever you want)
+
+KIGen (Keyme Inline Generator) is a tool for inline (and in-place)
```
KeyMe?

Jeff-Ciesielski replied:
* grabs the shame bell *

shonfeder commented on README.md:
```diff
@@ -0,0 +1,147 @@
+The start segment should follow the format: `# KIGEN_start foo arg1:val1`
+The end segment should follow the format: `# KIGEN_end`
+
+
```
rm extra line?

shonfeder commented on README.md:
```diff
@@ -0,0 +1,147 @@
+be supplied to the autogen module.  These are useful when you need to
+parameterize a single autogen module for slightly different usecases
+(i.e maybe you want to generate two different languages in the same
+module).  Any number of key value pairs may be provided
```
Missing a period.

shonfeder commented on README.md:
```diff
@@ -0,0 +1,147 @@
+
+Assume three files:
+
+A file to generate that contains the text:
```
Maybe give this a name too? Perhaps `bar.sh` (or whatever language you imagine this being?). Also, since this is mainly targeted at source code generation, maybe throw a line of simple code in the example, rather than just text?

shonfeder commented on README.md:
````diff
@@ -0,0 +1,147 @@
+And a content file (`foo.py`) containing:
+```
+def get_content(**kwargs):
+    return {'foo': 'World'}
````
Does it mean something special that the name of the dictionary key is the same as the name of the module? If not, probably good to make these different. If so, it would be helpful for me to have this meaning explained.

shonfeder commented on README.md:
```diff
@@ -0,0 +1,147 @@
+created (if it doesn't already exist) and the rendered files will be
+written in that directory.
+
+ex: `kigen -i path/to/file.py -i /path/to/file2.py -m mod/dir/one -m /mod/dir/2`
```
Would be helpful to make this example maybe match the hello world. Or perhaps provide one example that shows how you'd generate the hello world, and a second one showing use of multiple input files and module paths?

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+    pass
+
+
+def expand_template(template_str: str, content) -> str:
```
I'm curious about the inconsistent use of type hints in these functions...

Jeff-Ciesielski replied:
It was mostly to keep return types and incoming stuff straight in my head.  I'm very torn about including them TBH as I don't derive any use from their inclusion...

shonfeder replied:
imo, they can work well as documentation, to keep stuff clear in our heads. I can also help you get this passing type checking under mypy, if you'd like. And that should impart some use.

I guess the inconsistent use of type hints doesn't do any harm, but I do find it a bit strange. My inclination is to suggest you either use them consistently (perhaps in a stubfile, if you don't want the added density in your code), or remove them all together. But if you feel that the use as-is helps document, I'm OK with you leaving it.

Jeff-Ciesielski replied:
I think I'm going to opt for their removal.  Can always add them in (or add stubs).

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+                raise NestedBlockError(
+                    "Nested blocks are not supported. "
+                    "Detected a new block at line {} "
+                    "but the block startint at {} has not "
```
typo `startint`

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+    command = None
+    commentmarker = None
+    for idx, line in enumerate(file_data.splitlines()):
+        if START_MARKER in line:
```
iiuc, users of this tool have to be sure that the sequence of characters `KIGEN_start` never appears in any of their code (even as a subsequence). This seems a relatively safe bet, but if it's also easy for me to imagine a user referring to the kigen stuff in a comment like:

```
# Everything in the block starting with `KIGEN_start` will blah blah
```

which, I think, would break the rendering. 

I.e. as a naive user I would expect the generator to only for lines matching the pattern `comment_marker keyword`, but it looks like it will take any occurrence of the keywords and assume the preceding sequence is a comment marker. In the future, it would be could to clean this up (maybe just by being sensitive to the comment markers for target languages?), but for the time being, it would be a good safeguard to at least remark this in the documentation.

Jeff-Ciesielski replied:
Good thinking.

Jeff-Ciesielski replied:
I've addressed this with a comment.  I need to do some restructuring to make this possible.

shonfeder reviewed (commented):
My first sally. I still haven't gotten deep in to the code. What I've read so far seems very clear and well documented to me :+1:

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+    """
+    assert os.path.isdir(path)
+
+    files = [x for x in os.listdir(path)
```
Should this search recurse through directory and its children, so that we can kigen a whole project without having to pass in every directory individually?

Jeff-Ciesielski replied:
no, it's intended to only search specified directories.

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+        if full_package_name not in known_modules:
+            continue
+
+        if full_package_name not in sys.modules:
```
I'm a bit worries this could lead to confusing cases where modules fail to load due to trivial name collisions. E.g., looks like there's a standard module `encodings` in my `sys.modules`. So if I had a kigen module by this name, I wouldn't be able to load it, but I also wouldn't get any warnings. I'm not sure if this something you can or should address here, or if its just an ineradicable problem. One way to to work around it might be to take the module's file location in to account to disambiguate these cases...

Maybe this is something better left till we see whether it actually causes problems.

Jeff-Ciesielski replied:
I'll enhance error handling here.

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+
+
+def expansion_module_to_template(exp_mod: ExpansionModule) -> str:
+    """Reads the template portion of an ExpansionMOdule and returns a
```
Typoe `ExpansionMOdule`

Jeff-Ciesielski replied:
I <3 that there is a typo in your comment about my typo :D

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+    start_str = block_to_start_string(block)
+
+    try:
+        exp_mod = modules[block.command.function]
```
I'm a bit confused by this: it looks like we're finding a module by looking up a function, whereas I know that the dict `modules` has module names as keys. I assume, therefore, that `block.command.function` here must actually be the name of a module... but then I'm not sure why it's called function! I could puzzle this out by revisiting the data structures, but I wonder if the confusion on site here suggests an improvement that could be made to the naming of the data structures?

Jeff-Ciesielski replied:
The original name of 'modules' was 'functions' as the calling convention in the comments resembles a function + arguments (i.e. foo arg1:val1 arg2:val2).

I'll just change this name.

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+
+def recombine(chunks, blocks) -> str:
+    """Given a list of strings containing the non-autogen portions of a
+    file, and a list of strings containng rendered AutogenBlocks,
```
typo `containng`

shonfeder commented on kigen.py:
```diff
@@ -0,0 +1,401 @@
+    )
+    parser.add_argument(
+        '-o', '--output-dir',
+        action="store",
```
Switch to single quotation marks for consistency.  You win at life forever

shonfeder commented on test_collateral/bar.jinja2:
```diff
@@ -0,0 +1 @@
+Goodbye {{ bar }}
```
What does "collateral" mean here?

Jeff-Ciesielski replied:
data used for testing.

shonfeder commented on test_collateral/file_to_autogen.rb:
```diff
@@ -0,0 +1,6 @@
+Parts of this file are autogenerated
+# KIGEN_start foo
+# KIGEN_end
+Cool bean.
```
Just one single, lonesome cool bean.

Jeff-Ciesielski replied:
You can thank Dillon for this.

shonfeder reviewed (commented):
Finished a pass! I'm afraid I was likely skimming too shallow by the end, and I'll look over the later third once again to be sure.

knitHacker commented on README.md:
```diff
@@ -0,0 +1,147 @@
+
+KIGen is easy to use and the only requirement is that the target file
+supports inline comment markers (i.e. `#` or `//`, but not
+`/**/`...Sorry pedantic C89 folks)
```
**cough cough** you and stephan...

Jeff-Ciesielski committed:
Misc comment cleanup

Jeff-Ciesielski committed:
Add error handling due to module conflicts

Jeff-Ciesielski committed:
Change 'function' to 'module' in ModuleCmd

This name is confusing since it LOOKS like a function definition in
the kigen marker, but is in reality the name of a module that exposes
a 'get_content' method.

Jeff-Ciesielski committed:
Fix tests

Jeff-Ciesielski committed:
Add documentation about markers not appearing in code

Jeff-Ciesielski commented:
Alright, I think this is *finally* ready for final review.  Sorry about the delay on this.

I think I'd like to try making this the basis for XBVC at some point since they fill a similar niche, but this is more generalized.

knitHacker reviewed (approved):
LGTM
